packages/runtime/container-runtime/src/test/garbageCollection.spec.ts

```diff
@@ -829,36 +846,152 @@ describe("Garbage Collection Tests", () => {
+            nodes.forEach((nodeId) => {
+                garbageCollector.nodeUpdated(nodeId, "Changed", Date.now(), testPkgPath);
+                garbageCollector.nodeUpdated(nodeId, "Loaded", Date.now(), testPkgPath);
+            });
+        }
+
+        const mainContainerTests = (
+            timeout: number,
+            loadedEventName: string,
+            sweepReadyUsageDetectionEnabled: boolean,
+            snapshotCacheExpiryMs: number,
+        ) => {
+            let lastCloseErrorType: string = "N/A";
+
+            const createGCOverride = (
+                baseSnapshot?: ISnapshotTree,
+                gcBlobsMap?: Map<string, IGarbageCollectionState | IGarbageCollectionDetailsBase>,
+            ) => {
+                const gc2 = createGarbageCollector({ baseSnapshot, snapshotCacheExpiryMs }, gcBlobsMap, (error) => { lastCloseErrorType = error?.errorType ?? "NONE"; });
+                (gc2 as any).isSummarizerClient = false; // Main Container
+                return gc2;
+            };
+
+            it("generates events and may close container when SweepReady nodes are used", async () => {
+                // Create GC state where node 3's unreferenced time was > timeout ms ago.
+                // This is important since we shouldn't run GC on the main container,
+                // but rather load from a snapshot in which SweepReady state is already reached.
+
+                // Create a snapshot tree to be used as the GC snapshot tree.
+                const gcSnapshotTree = getDummySnapshotTree();
+                const gcBlobId = "root";
+                // Add a GC blob with key that start with `gcBlobPrefix` to the GC snapshot tree. The blob Id for this
+                // is generated by server in real scenarios but we use a static id here for testing.
+                gcSnapshotTree.blobs[`${gcBlobPrefix}_${gcBlobId}`] = gcBlobId;
+
+                // Create a base snapshot that contains the GC snapshot tree.
+                const baseSnapshot = getDummySnapshotTree();
+                baseSnapshot.trees[gcTreeKey] = gcSnapshotTree;
+
+                // Create GC state with node 3 expired. This will be returned when the garbage collector asks
+                // for the GC blob with `gcBlobId`.
+                const gcState: IGarbageCollectionState = { gcNodes: {} };
+                const node3Data: IGarbageCollectionNodeData = {
+                    outboundRoutes: [],
+                    unreferencedTimestampMs: Date.now() - (timeout + 100),
+                };
+                gcState.gcNodes[nodes[3]] = node3Data;
+
+                const gcBlobMap: Map<string, IGarbageCollectionState> = new Map([[gcBlobId, gcState]]);
+                const garbageCollector = createGCOverride(baseSnapshot, gcBlobMap);
+
+                // Trigger loading GC data from base snapshot - but don't call GC since that's not what happens in real flow
+                await (garbageCollector as any).initializeBaseStateP;
```

Comment:
I wish there was a way to avoid this cast to any :-(

Reply:
Could we call `setConnectionState` that my recent PR added instead? That is a sync function though, so I am not sure how to wait for the initialization to complete.

Reply:
Eh I'm a lot less worried about `any` in unit tests because if you get it wrong the test will fail.  In production code you can get it wrong but not know because you won't be there when the user runs the code and it's broken.

Reply:
(and no `setConnectionState` won't help since it's not async, as you say)